docs/adr/0007-preventing-version-mismatches.md

```diff
@@ -0,0 +1,269 @@
+code in `build.rs`, we could use a similar system:
+
+- The Rust crate adds a dependency to each bindings generator they want to
+  use (`uniffi-bindgen-kotlin = 1.10.1`, `uniffi-bindgen-swift = 1.4.0`,
+  etc).
+- Each bindings generator defines a `generate_bindings()` function, which
+  generates the bindings and writes them to the `target/[language]` directory.
+- We call this in `build.rs`:
+
+```
+fn main() {
+  uniffi_bindgen_language1.generate_bindings();
+  uniffi_bindgen_language2.generate_bindings();
+  ...
+}
+```
+
+- The build scripts would then copy the bindings from that directory to the
+  final location.
+
+Pros:
+  - Doesn't rely on a system-wide install of `uniffi-bindgen`.
+  - Bindgen tool versions are specified in the `Cargo.toml`, alongside the
+    `uniffi` version.
+
+Cons:
+  - Forces a specific order to the build process.  If you try building the
+    bindings before the Rust component, it would fail since the source files
+    haven't been generated yet (or worse, succeed with stale code).
+  - Hard to make this work with downstream crates.  For the desktop JS example,
+    the app-services components would need to include a call to
+    `uniffi_bindgen_gecko_js.generate_bindings()` in the `build.rs` file which
+    is not ideal.  Also, it would generate Kotlin/Swift bindings, which aren't
+    needed.  I think the solution would be to create a feature gate for each
+    bindings language that we could enable/disable, but this seems overly
+    complex.  Even this solution wouldn't work for some 3rd-party cases, for
+    example a 3rd-party that wants to use their home-grown bindings generator.
+
+### **[Option 2] Tell users to create embedded bindgen crates**
+
+We could make the `application-services` solution the recommended system:
+
+- Require each bindgen tool to expose a `clap_command()` function that returns
+  a `clap::Command` instance.
+- Create a `uniffi_bindgen::run_bindgen_app` function that executes a CLI app
+  from a list of `clap::Command` instances from bindgen tools.  Each one of
+  these commands would become a clap subcommand.
+- Tell UniFFI users to create a new crate in their workspace with a bindgen
+  binary.  The code would look something like this:
+
+```
+fn main() {
+    uniffi_bindgen::run_bindgen_app(vec![
+       uniffi_bindgen_language1::clap_command(),
+       uniffi_bindgen_language2::clap_command(),
+       uniffi_bindgen_language3::clap_command(),
+       ...
+    ])
+}
+```
+
+Pros:
+  - Doesn't rely on a system-wide install of `uniffi-bindgen`.
+  - Bindgen tool versions are specified in the `Cargo.toml`, alongside the
+    `uniffi` version.
+  - Doesn't require building the main library to generate bindings.
+  - Supports in the downstream crate scenario.  If we take the moz-central
+    example from above, it would create it's own bindgen crate that only
+    generated `gecko-js` bindings, and that crate could generate bindings for
+    the app-services UDL files.
+
```

Comment:
can we formalize the app-services code inside the uniffi repo in some way? We'd need to get creative, but feature gates might mean uniffi helps support all "known" bindings, even for bindings not in our repo, or something hand-wavey.

(I guess I'm thinking that some of the cons below might not necessarily be as bad as they look on paper - eg, #1190 has a similar work-around. OTOH, maybe the burden is small enough once a consumer becomes "sophisticated" enough, but it might be a large hurdle for new uniffi users who are just investigating it and are you to commit)